RavenDB-15742 : OLAP ETL

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-15742

aviv86 committed:
RavenDB-15742 : initial work on OLAP ETL

ayende commented on src/Raven.Client/Documents/Operations/ETL/EtlType.cs:
```diff
@@ -3,6 +3,7 @@
         Raven,
-        Sql
+        Sql,
+        S3
```
I don't like the S3 designation here. Maybe Parquet, instead?

ayende commented on src/Raven.Client/Documents/Operations/ETL/S3ConnectionString.cs:
```diff
@@ -0,0 +1,29 @@
+        protected override void ValidateImpl(ref List<string> errors)
+        {
+            if (S3Settings == null)
+                errors.Add($"{nameof(S3Settings)} cannot be null");
```
we have that in the backup already, no?
Can we merge the impl?

ayende commented on src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:
```diff
@@ -14,6 +14,7 @@ public enum OngoingTaskType
         Replication,
         RavenEtl,
         SqlEtl,
+        S3Etl,
```
ParquetEtl ?

ayende commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -212,6 +220,13 @@ private void HandleChangesSubscriptions()
 
                         break;
+                    case EtlType.S3:
+                        s3Config = config as S3EtlConfiguration;
```
Need to think about how we can integrate this with the user. Can we use a script to get the details, as well?
That way they can generate that on the fly. Like we do for backups.

ayende commented on src/Raven.Server/Documents/ETL/Providers/S3/RowGroups.cs:
```diff
@@ -0,0 +1,149 @@
+        public void Add(ToS3Item item)
+        {
+            var group = GetCurrentGroup(item);
+            var propCount = item.Properties.Count;
```
What happens when different items have different number of fields?
What happens if the order of the field isn't fixed?

ayende commented on src/Raven.Server/Documents/ETL/Providers/S3/RowGroups.cs:
```diff
@@ -0,0 +1,149 @@
+
+        private RowGroup GetCurrentGroup(ToS3Item item)
+        {
+            if (Groups.Count == 0 || Groups[^1].Count == MaxItemsPerGroup)
```
That looks great

ayende commented on src/Raven.Server/Documents/ETL/Providers/S3/S3Etl.cs:
```diff
@@ -0,0 +1,171 @@
+        private string GetTempPath()
+        {
+            var dir = Configuration.TempDirectoryPath ?? Path.GetTempPath();
+            var fileName = $"{Guid.NewGuid()}.parquet"; 
```
Should have - db name, etl name, date as well.

ayende commented on src/Raven.Server/Documents/ETL/Providers/S3/S3Etl.cs:
```diff
@@ -0,0 +1,171 @@
+            return new S3DocumentTransformer(Transformation, Database, context, Configuration);
+        }
+
+        protected override int LoadInternal(IEnumerable<RowGroups> records, DocumentsOperationContext context)
```
What creates the batches here?

ayende commented on test/SlowTests/Server/Documents/ETL/S3EtlTests.cs:
```diff
@@ -0,0 +1,170 @@
+var month = orderDate.getMonth();
+var key = new Date(year, month);
+
+loadToOrders(key,
```
what is the meaning of `Orders` here?

ayende commented on test/SlowTests/Server/Documents/ETL/S3EtlTests.cs:
```diff
@@ -0,0 +1,170 @@
+                {
+                    new Transformation
+                    {
+                        Name = "MonthlyOrders",
```
What is the meaning of `MonthlyOrders` here?

ayende commented on src/Raven.Server/Documents/ETL/Providers/S3/S3Etl.cs:
```diff
@@ -0,0 +1,171 @@
+        {
+            Metrics = S3Metrics;
+
+            _timer = new Timer(_ => _waitForChanges.Set(), null, TimeSpan.Zero, configuration.ETLFrequency);
```
If I understand this properly, we are handling the delay of this kind of ETL using the timer?
So we'll start running on startup, then every `ETLFrequency` ?

karmeli87 replied:
This is a good point. 
We need to persist last time it was run and adjust the timer, so we will not start on each server restart.

ayende replied:
Take into account that this may _move_ between nodes as well, so we have to account for that as well.

aviv86 committed:
RavenDB-15742 : persist last batch time and adjust the timer, so we will not start on each server restart

aviv86 committed:
RavenDB-15742 : RowGroups - change implementation to not rely on the order of fields, and allow to have items with different number of fields

aviv86 committed:
RavenDB-15742 : add more tests

aviv86 committed:
RavenDB-15742 : row groups - change dictionary from DataField to DataType in order to support unknown field types, fix bugs, add tests

aviv86 committed:
RavenDB-15742 ParquetEtl : support getting S3 settings from script and other minor additions

aviv86 committed:
RavenDB-15742 : minor refactoring, report statistics, add test

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/EtlProcessState.cs:
```diff
@@ -27,7 +27,8 @@ public EtlProcessState()
         // we mark the time-series of the document as skipped so when we load the document we will load all its time-series with it
         public HashSet<string> SkippedTimeSeriesDocs { get; set; }
-        
+        public long LastBatchTime { get; set; }
```
Any reason why is it `long` instead of `DateTime`?

arekpalinski replied:
Still same question ^

aviv86 replied:
changed to `DateTime?`

arekpalinski replied:
Yes, let's go with `DateTime?`

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/ParquetEtlConfiguration.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        public string TempDirectoryPath { get; set; }
+
+        public TimeSpan ETLFrequency { get; set; }
```
Maybe `RunFrequency`? We are already in "ETL" context

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/ParquetEtlConnectionString.cs:
```diff
@@ -0,0 +1,58 @@
+                        errors.Add($"{nameof(S3Settings)} has no setting");
+                    return;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
```suggestion
                    throw new ArgumentOutOfRangeException(nameof(Destination), Destination, "Unsupported destination type");
```

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/RavenConnectionString.cs:
```diff
@@ -12,7 +12,7 @@ public class RavenConnectionString : ConnectionString
         public override ConnectionStringType Type => ConnectionStringType.Raven;
 
-        protected override void ValidateImpl(ref List<string> errors)
+        public override void ValidateImpl(ref List<string> errors)
```
Why public? For public usage we have ` public bool Validate(ref List<string> errors)`

arekpalinski commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1310,6 +1310,11 @@ public void AssertCanAddSqlEtl()
+        
+        public void AssertCanAddParquetEtl()
+        {
+            // todo
```
Please create a dedicated issue for handling this by @grisha-kotler

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -83,14 +86,15 @@ private void OnProcessRemoved(EtlProcess process)
                 RavenDestinations = _databaseRecord.RavenEtls;
                 SqlDestinations = _databaseRecord.SqlEtls;
-
+                S3Destinations = _databaseRecord.ParquetEtls;
```
Why `S3Destinations` as the name? S3 is potentially just one of the destinations for Parquet ELT, isn't it?

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -213,7 +221,13 @@ private void HandleChangesSubscriptions()
                         break;
-
+                    case EtlType.Parquet:
+                        s3Config = config as ParquetEtlConfiguration;
```
Same here - why `s3Config`, not `parquetConfig`?

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlProcess.cs:
```diff
@@ -709,11 +711,13 @@ public void Run()
-                    if (didWork)
+                    if (didWork || ShouldUpdateOnLastBatch)
                     {
+                        var batchTimeMilliseconds = didWork ? 0 : startTime.Ticks / 10_000;
```
Why dealing with ticks here? What's the pros instead of using `DateTime`?

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlProcess.cs:
```diff
@@ -709,11 +711,13 @@ public void Run()
                     }
 
-                    if (didWork)
+                    if (didWork || ShouldUpdateOnLastBatch)
```
I don't follow why we're _always_ updating the ETL state here for ParquetETL? Why `didWork ` was not sufficient?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+                    ? string.Empty
+                    : $"{s3Settings.RemoteFolderName}/";
+
+                var key = Path.Combine(prefix, path);
```
Didn't we want to add some ETL process specific prefix? In order to be able to distinguish between two Parquete ETL file uploaded to the same folder on S3?

ayende replied:
Yes, some prefix (used defined?) is required

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetDocumentTransformer.cs:
```diff
@@ -0,0 +1,146 @@
+
+        protected override string[] LoadToDestinations { get; }
+
+        protected override void LoadToFunction(string tableName, ScriptRunnerResult colsAsObject)
```
I wonder what is the meaning of "table" used as parameter name in context of Parquet ETL?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetDocumentTransformer.cs:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        protected override void AddLoadedAttachment(JsValue reference, string name, Attachment attachment)
+        {
```
I assume attachments aren't gonna be supported?

ayende replied:
Not planning to

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetDocumentTransformer.cs:
```diff
@@ -0,0 +1,146 @@
+
+        protected override void AddLoadedCounter(JsValue reference, string name, long value)
+        {
+            throw new NotSupportedException("Counters aren't supported by SQL ETL");
```
by SQL ETL -> by Parquet ETL

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetDocumentTransformer.cs:
```diff
@@ -0,0 +1,146 @@
+
+        protected override void AddLoadedTimeSeries(JsValue reference, string name, IEnumerable<SingleResult> entries)
+        {
+            throw new NotSupportedException("Time series aren't supported by SQL ETL");
```
by SQL ETL -> by Parquet ETL

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+            var etlFrequency = _configuration.ETLFrequency;
+            var dueTime = GetDueTime(etlFrequency);
+
+            _timer = new Timer(_ => _waitForChanges.Set(), null, dueTime, etlFrequency);
```
Does it needs to be controlled directly by the timer? This mean's that you're forcing to run ETL even if there isn't anything do to.

I was thinking that that actual implementation will be more like this:

- wait for document changes - same as in current impl
- once you get a signal about changes then:
```
if (now - time_since_last_run > ETLFrequency) 
{
  RunParquetETL() 
}
else  
{ 
  WaitAppropriateTimeToReachETLFrequency();
  continue; // to actually run it now
}
```

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+        protected override IEnumerator<ToParquetItem> ConvertTombstonesEnumerator(DocumentsOperationContext context, IEnumerator<Tombstone> tombstones, string collection, bool trackAttachments)
+        {
+            return EmptyEnumerator; // todo
+            throw new NotSupportedException("Tombstones aren't supported by S3 ETL");
```
S3 ETL ?

arekpalinski replied:
In other places as well

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+        private string GetPath(RowGroups group, out string remotePath)
+        {
+            var dir = Configuration.TempDirectoryPath ?? Path.GetTempPath();
+            var fileName = $"{Database.Name}_{Guid.NewGuid()}.parquet";
```
I think it would make sense to include ETL process name too

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+                        case DataType.Float:
+                        case DataType.Double:
+                        case DataType.Decimal:
+                            array = ((List<double>)data).ToArray();
```
Can we use `ArrayPool` to avoid allocating so many arrays?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+            return count;
+        }
+
+        private void WriteGroup(ParquetWriter parquetWriter, RowGroup group, RowGroups rowGroups)
```
Wouldn't it make sense to move this method and its implementation to `ParquetWriter` as well?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+                    //UploadToS3(s3Settings, stream, path);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Please provide param name and value to ArgumentOutOfRangeException

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetEtl.cs:
```diff
@@ -0,0 +1,260 @@
+
+                var key = Path.Combine(prefix, path);
+
+                client.PutObject(key, stream, new Dictionary<string, string>
```
I think we need to expose the progress of upload to user sso they will know that ETL process isn't stuck.

arekpalinski replied:
Also I think we should let user know that if the upload takes more time that `ETLFrequency` - so he might need to adjust that setting.

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetDocumentTransformer.cs:
```diff
@@ -0,0 +1,146 @@
+
+            if (_tables.TryGetValue(name, out var table) == false)
+            {
+                _tables[name] = table = new RowGroups(tableName, key);
```
This dictionary can be potentially very large because it keeps _all_ records in memory, right? As I understand the batches in Parquet ETL will be significantly larger because they aggregate data modified in the last few hours (depending on `ETLFrequency` configuration).

Do we have a way to write the result of LoadTo() immediately to the temporary parquet file? So we would not need keep it in memory

arekpalinski replied:
Another thing that is concerning is the duration of the transaction. We're processing entire batch in single transaction right?

ayende replied:
I would like to build them on the fly, yes.
We can write multiple RowGroups, I think

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/RowGroups.cs:
```diff
@@ -0,0 +1,256 @@
+
+        private Dictionary<string, DataField> _fields;
+
+        private const int MaxItemsPerGroup = 50_000;
```
Shouldn't it be configurable for a user? (advanced setting)

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/ParquetDocumentTransformer.cs:
```diff
@@ -0,0 +1,146 @@
+        public override void Transform(ToParquetItem item, EtlStatsScope stats, EtlProcessState state)
+        {
+            _stats = stats;
+            if (item.IsDelete)
```
Given that we do nothing on delete I then we can not iterate over tombstones at all, right?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/RowGroups.cs:
```diff
@@ -0,0 +1,256 @@
+                        data ??= new List<bool>();
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
```
please provide parameters to ArgumentOutOfRangeException

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Parquet/RowGroups.cs:
```diff
@@ -0,0 +1,256 @@
+                        continue;
+                    case BlittableJsonToken.Integer:
+                        propType = DataType.Int64;
+                        data ??= new List<long>();
```
To make sue -  is everything a list here because parquet format is column based?

arekpalinski commented on src/Raven.Server/ServerWide/Commands/ConnectionStrings/RemoveConnectionStringCommand.cs:
```diff
@@ -61,4 +61,22 @@ public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
         }
     }
+
+    public class RemoveOlapConnectionStringCommand : RemoveConnectionStringCommand<ParquetEtlConnectionString>
```
Is it used anywhere? It's the first time I see "OLAP"

arekpalinski reviewed (commented):
After reviewing entire PR I have some doubts about the implementation. My concern is that the code is super "parquet" oriented - starting from small things like enum `EtlType.Parquet` and ending at the main class `ParquetEtl`.

Actually I thought that it's gonna be more "OLAP" oriented and parquet format will be just one of the possible output providers (the only one - in the first implementation). That could give us some flexibility in adding more formats in the future.

Same approach when it comes to the upload - right now just S3 is supported but in the future we might have more. So S3 isn't Parquet ETL specific.

aviv86 committed:
RavenDB-15742 : upload to S3 + tests

aviv86 committed:
RavenDB-15742 : more tests

ayende commented:
The idea is that we are going to be generating Parquet files, the choices are Parquet, Avro or ORC, pretty much.
Avro is row based, and not likely fit this need. Parquet is better supported than ORC.

The name OLAP ETL can be used, yes, and we should have a single option file type. But in essence, I don't think we'll be adding
any additional formats easily. 

And the upload process *must* support anything that the backup does, and it should use the exact same code.

aviv86 committed:
RavenDB-15742 : introduce OlapTransformedItems abstraction in order to support more file formats in the future

aviv86 committed:
RavenDB-15742 : change ETLProccess.Transform() and EtlTransformer.GetTransformedResults() to return IEnumerable instead of List (so we can avoid those ToList calls)

aviv86 committed:
RavenDB-15742 : refactoring +  address some PR comments

aviv86 committed:
RavenDB-15742 : add support for DateTimeOffset and TimeSpan data types, error handling

aviv86 committed:
RavenDB-15742 : when writing parquet row groups, reuse array instances instead of calling ToArray

aviv86 committed:
RavenDB-15742 : EtlProcess.cs :
- revert changes in Run()
- remove ShouldUpdateOnLastBatch
- introduce AfterAllBatchesCompleted() (implemented only by OLAP etl)
- minor refactoring

aviv86 committed:
RavenDB-15742 : add _id and _lastModifiedTicks fields to parquet output

aviv86 committed:
RavenDB-15742 : introduce partition filed name and modify remote paths to be hive compatible (e.g. 'dt=2020-03-01-00-00'), apply changes to tests

aviv86 committed:
RavenDB-15742 : allow to modify document id column name & partition column name (for each output 'table')

aviv86 committed:
RavenDB-15742 : add more tests
- test with output to multiple tables
- test modifying id and partition column names
- test with partitioning by day and by hour

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/OLAP/OlapEtlConfiguration.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public List<OlapEtlTable> OlapTables { get; set; }
+
+        public override string GetDestination() => ConnectionStringName;
```
Just a name of connection string isn't very meaningful. Maybe let's extract S3 details here - e.g. bucket name. That's the real destination of OLAP ETL files

ayende replied:
Note that this ETL isn't about S3, it can go to SFTP, for example.

arekpalinski replied:
Exactly - we should pull the details about connection from the connection string - not returning the connection string name. 

We already do it for SQL and Raven ETLs via `Connection` property

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/OLAP/OlapEtlConfiguration.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public override bool UsingEncryptedCommunicationChannel()
+        {
+            throw new NotImplementedException();
```
We need to have it handled when outputting data from encrypted db

ayende replied:
For S3, Azure, GCP storage - can assume this is safe.
Local path, also safe.
For FTP - abort
SFTP - Allow

arekpalinski replied:
Still `NotImplementedException`?

aviv86 replied:
created a separate issue to handle that https://issues.hibernatingrhinos.com/issue/RavenDB-16627

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlapDocumentTransformer.cs:
```diff
@@ -0,0 +1,169 @@
+            if (args[1].IsObject() == false)
+                ThrowInvalidScriptMethodCall($"loadTo{name}(key, obj) argument 'obj' must be an object");
+
+            var key = args[0].AsDate().ToDateTime().ToString(DateFormat); //todo
```
What about this `todo`?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlaptEtl.cs:
```diff
@@ -0,0 +1,202 @@
+
+        protected override IEnumerator<ToOlapItem> ConvertTombstonesEnumerator(DocumentsOperationContext context, IEnumerator<Tombstone> tombstones, string collection, bool trackAttachments)
+        {
+            return EmptyEnumerator;
```
Tombstones are just ignored, right? This is important to write about that explicitly in the documentation

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlaptEtl.cs:
```diff
@@ -0,0 +1,202 @@
+
+        private void UploadToS3(S3Settings s3Settings, Stream stream, string key)
+        {
+            using (var client = new RavenAwsS3Client(s3Settings, progress: null, Logger, CancellationToken))
```
`progress: null` - can we expose it to the user? We already know how many files has been uploaded, and from here we might know the upload progress of current file, right?

If this is feasible please create a dedicates issue to handle that

ayende replied:
Yes, that is a good point to have, they might be _large_.
Also need to make sure we handle resuming uploads.

aviv86 replied:
https://issues.hibernatingrhinos.com/issue/RavenDB-16341

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlaptEtl.cs:
```diff
@@ -0,0 +1,202 @@
+
+        private string GetPath(OlapTransformedItems transformed, out string remotePath)
+        {
+            var fileName = $"{Database.Name}__{Guid.NewGuid()}.{transformed.Format}";
```
Double `_` characters in a file name?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/ParquetTransformedItems.cs:
```diff
@@ -0,0 +1,445 @@
+                ? DefaultIdColumn
+                : idColumn;
+
+            Prefix = $"{CollectionName}/{partitionColumn}%3D{PartitionKey}";
```
What is `%3D`? Why it needs to be provided this way? Please create a dedicated const with proper name for that

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/ParquetTransformedItems.cs:
```diff
@@ -0,0 +1,445 @@
+
+        private Dictionary<string, DataField> _fields;
+
+        private const int MaxItemsPerGroup = 50_000;
```
Didn't we want to make it configurable (as an advanced option)?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/ParquetTransformedItems.cs:
```diff
@@ -0,0 +1,445 @@
+                    case BlittableJsonToken.StartArray:
+                    case BlittableJsonToken.CompressedString:
+                    case BlittableJsonToken.EmbeddedBlittable:
+                        // todo
```
What `todo` here?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/ParquetTransformedItems.cs:
```diff
@@ -0,0 +1,445 @@
+                        data ??= new List<long>();
+                        break;
+                    case BlittableJsonToken.LazyNumber:
+                        // todo
```
What `todo` here?

arekpalinski commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -144,6 +144,9 @@
     <PackageReference Include="NuGet.Protocol" Version="5.8.1" />
     <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="2.19.101" />
     <PackageReference Include="Portable.BouncyCastle" Version="1.8.9" />
+    <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="2.19.100" />
```
We already have it referenced - see above:

`<PackageReference Include="Oracle.ManagedDataAccess.Core" Version="2.19.101" />`

arekpalinski commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -144,6 +144,9 @@
     <PackageReference Include="Portable.BouncyCastle" Version="1.8.9" />
+    <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="2.19.100" />
+    <PackageReference Include="Parquet.Net" Version="3.8.4" />
+    <PackageReference Include="Portable.BouncyCastle" Version="1.8.8" />
```
We already have it referenced - see above:

`<PackageReference Include="Portable.BouncyCastle" Version="1.8.9" />`

aviv86 committed:
RavenDB-15742 : better integration with EtlProcess and EtlLoader, basic client work, test

aviv86 committed:
RavenDB-15742 : change from S3ETL to ParquetETL

aviv86 committed:
RavenDB-15742 : add local setting to ParquetEtlConfiguration and modify tests

aviv86 committed:
RavenDB-15742 : rename ParquetEtl to OlapEtl

aviv86 committed:
RavenDB-15742 : fix failing ETL tests

aviv86 committed:
RavenDB-15742 : address some PR comments

aviv86 committed:
RavenDB-15742 : don't keep more than one row group in memory at a time, write/append row group to file when we reach the limit

aviv86 committed:
RavenDB-15742 : refactoring : Load - use exact backup code

aviv86 committed:
RavenDB-15742 : make MaxNumberOfItemsInRowGroup configurable, add test, minor refactoring

aviv86 committed:
RavenDB-15742 : minor fixes

Danielle9897 committed:
RavenDB-16310 OLAP ETL : Edit Task View

aviv86 committed:
RavenDB-15742 : tmp changes

ml054 commented on src/Raven.Studio/typescript/commands/database/settings/getConnectionStringInfoCommand.ts:
```diff
@@ -28,6 +28,10 @@ class getConnectionStringInfoCommand extends commandBase {
     }
+
+    static forOlapEtl(db: database, connectionStringName: string) {
+        return new getConnectionStringInfoCommand(db, "Olap", connectionStringName);
```
OLAP

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskOlapEtlTransformationModel.ts:
```diff
@@ -0,0 +1,91 @@
+    toDto(): Raven.Client.Documents.Operations.ETL.Transformation {
+        return {
+            ApplyToAllDocuments: false,
+            Collections: [this.collection()],
```
why we support single collection? see: https://github.com/ravendb/ravendb/blob/c329d57f62a480d9b66e0bcc5c110b286d676c3b/src/Raven.Studio/typescript/models/database/tasks/ongoingTaskRavenEtlTransformationModel.ts#L59

Danielle9897 replied:
done, now support multiple

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/periodicBackup/backupConfiguration.ts:
```diff
@@ -1,6 +1,7 @@
 ﻿/// <reference path="../../../../../typings/tsd.d.ts"/>
 import localSettings = require("models/database/tasks/periodicBackup/localSettings");
-import s3Settings = require("models/database/tasks/periodicBackup/s3Settings");
+// import s3Settings = require("models/database/tasks/periodicBackup/s3Settings");
```
remove

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/smugglerDatabaseRecord.ts:
```diff
@@ -20,8 +20,10 @@ class smugglerDatabaseRecord {
     includeExternalReplications = ko.observable<boolean>(true);
     includeRavenConnectionStrings = ko.observable<boolean>(true);
     includeSqlConnectionStrings = ko.observable<boolean>(true);
+    includeOlapConnectionStrings = ko.observable<boolean>(true);
```
Is server side ready?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/settings/connectionStrings.ts:
```diff
@@ -195,47 +233,86 @@ class connectionStrings extends viewModelBase {
     }
     
+    private onOlapEtl() {
+        this.editedOlapEtlConnectionString().s3Settings().bucketName.subscribe(() => this.clearTestResult());
```
this is ugly. Can't we move that logic to s3settings object?

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/createOngoingTask.html:
```diff
@@ -42,6 +42,12 @@ <h5>ETL</h5>
                             <i class="icon-sql-etl"></i><br />SQL ETL
                         </a>
                     </div>
+                    <!-- todo... olap-icon...-->
```
icon

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/wwwroot/App/views/partial/backupDestinationLocal.html:
```diff
@@ -0,0 +1,58 @@
+                            <!-- ko ifnot: Error -->
+                            Free Disk Space:
+                            <span data-bind="if: FreeSpaceHumane">
+                                                        <strong data-bind="text: FreeSpaceHumane"></strong>
```
indent here is wrong?

ml054 commented on src/Raven.Studio/wwwroot/App/views/shell/about.html:
```diff
@@ -149,6 +149,7 @@ <h4>Backups</h4>
                     <div class="name">Encrypted backups</div><div class="value"><i data-bind="attr: { class: licenseAttribute('HasEncryptedBackups') }"></i></div>
                     <h4>ETL</h4>
                     <div class="name">SQL ETL</div><div class="value"><i data-bind="attr: { class: licenseAttribute('HasSqlEtl') }"></i></div>
+                    <div class="name">OLAP ETL</div><div class="value"><i data-bind="attr: { class: licenseAttribute('HasOlapEtl') }"></i></div>
```
we need to make sure that will be final license attribute name

Danielle9897 replied:
@grisha-kotler ?

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskOlapEtlListModel.ts:
```diff
@@ -0,0 +1,36 @@
+        this.initializeObservables();
+
+        this.connectionStringsUrl = appUrl.forConnectionStrings(activeDatabaseTracker.default.database(), "olap", this.connectionStringName());
+        console.log("url === " + this.connectionStringsUrl);
```
remove

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskOlapEtlTransformationModel.ts:
```diff
@@ -0,0 +1,91 @@
+    update(dto: Raven.Client.Documents.Operations.ETL.Transformation, isNew: boolean, resetScript: boolean) {
+        this.name(dto.Name);
+        this.script(dto.Script); 
+        this.collection(dto.Collections[0]);
```
why ?

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/periodicBackup/localSettings.ts:
```diff
@@ -2,18 +2,23 @@
-
-    constructor(dto: Raven.Client.Documents.Operations.Backups.LocalSettings) {
+    
+    container = ko.observable<container>();
```
what's the purpose of that?

ml054 replied:
can we avoid having that?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/settings/connectionStrings.ts:
```diff
@@ -166,21 +199,26 @@ class connectionStrings extends viewModelBase {
-            buttons: ["Cancel", "Delete"],
-            html: true
-        })
+        const stringType = connectionStringtype === "Raven" ? "RavenDB" :  connectionStringtype === "Sql" ? "SQL" : "Olap";
```
I've seen this login in some other place - I think it is time to move that to some static method with maps that .

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/settings/connectionStrings.ts:
```diff
@@ -264,7 +345,7 @@ class connectionStrings extends viewModelBase {
         if (sqlConnectionString) {
             if (this.isValid(sqlConnectionString.testConnectionValidationGroup)) {
-                eventsCollector.default.reportEvent("ravenDB-SQL-connection-string", "test-connection");
+                eventsCollector.default.reportEvent("SQL-connection-string", "test-connection");
```
please avoid changing that - we already have events in GA - we want to correlate them!

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/settings/connectionStrings.ts:
```diff
@@ -279,7 +360,7 @@ class connectionStrings extends viewModelBase {
         this.clearTestResult();
         const ravenConnectionString = this.editedRavenEtlConnectionString();
-        eventsCollector.default.reportEvent("ravenDB-ETL-connection-string", "test-connection");
+        eventsCollector.default.reportEvent("ravenDB-connection-string", "test-connection");
```
same here.

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editOlapEtlTask.ts:
```diff
@@ -0,0 +1,583 @@
+import jsonUtil = require("common/jsonUtil");
+import popoverUtils = require("common/popoverUtils");
+import tasksCommonContent = require("models/database/tasks/tasksCommonContent");
+import backupSettings = require("../../../models/database/tasks/periodicBackup/backupSettings");
```
update import

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editOlapEtlTask.ts:
```diff
@@ -0,0 +1,583 @@
+                .execute()
+                .done((result: Raven.Client.Documents.Operations.OngoingTasks.OngoingTaskOlapEtlDetails) => {
+                    this.editedOlapEtl(new ongoingTaskOlapEtlEditModel(result));
+                    this.showAdvancedOptions(!!result.Configuration.MaxNumberOfItemsInRowGroup);
```
let's move that logic to separate method - it will be helpful if we add extra conditions in future.

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editOlapEtlTask.ts:
```diff
@@ -0,0 +1,583 @@
+                content: `<small>Frequency at which this task will be triggered</small>`
+            });
+
+        popoverUtils.longWithHover($(".max-items-in-row"),
```
this is under advanced so may not be visible by default (please make sure tooltip is working when you defining new OLAP)

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editOlapEtlTask.ts:
```diff
@@ -0,0 +1,583 @@
+
+        overwriteAction.done(() => {
+            this.editedOlapEtl().olapTables.sort((a, b) => a.tableName().toLowerCase().localeCompare(b.tableName().toLowerCase()));
+            if (overwriteAction) { this.editedOlapTableSandbox(null); } 
```
format code

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/settings/connectionStrings.html:
```diff
@@ -67,23 +67,56 @@ <h4 class="text-elipsis">
                                 <div class="task">
-                                    <a data-bind="attr: { 'href': $root.taskEditLink($data.taskId, $parent.connectionStringName()) }" title="Click to edit task.."><span class="text-info small" data-bind="text: $data.taskName"></span></a>
+                                    <a data-bind="attr: { 'href': $root.taskEditLink($data.taskId, $parent.connectionStringName()) }"
+                                       title="Click to edit task.."><span class="text-info small" data-bind="text: $data.taskName"></span></a>
```
triple dots?

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/settings/connectionStrings.html:
```diff
@@ -94,7 +127,20 @@ <h3 class="flex-grow" data-bind="visible: editedSqlEtlConnectionString">SQL Conn
                                 <div class="task"> 
-                                    <a data-bind="attr: { 'href': $root.taskEditLink($data.taskId, $parent.connectionStringName()) }" title="Click to edit task.."><span class="text-info small" data-bind="text: $data.taskName"></span></a>
+                                    <a data-bind="attr: { 'href': $root.taskEditLink($data.taskId, $parent.connectionStringName()) }"
+                                       title="Click to edit task.."><span class="text-info small" data-bind="text: $data.taskName"></span></a>
```
triple dots?

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/settings/connectionStrings.html:
```diff
@@ -94,7 +127,20 @@ <h3 class="flex-grow" data-bind="visible: editedSqlEtlConnectionString">SQL Conn
+                            <div data-bind="foreach: tasksThatAreUsingThisConnectionString">
+                                <div class="task">
+                                    <a data-bind="attr: { 'href': $root.taskEditLink($data.taskId, $parent.connectionStringName()) }"
+                                       title="Click to edit task.."><span class="text-info small" data-bind="text: $data.taskName"></span></a>
```
triple dots? + $data.taskName -> taskName

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editOlapEtlTask.html:
```diff
@@ -0,0 +1,294 @@
+                                    <label for="keepFiles">Keep files on disk</label>
+                                </div>
+                            </div>
+                            <div class="form-group">
```
I wonder if this should go to advanced?

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editOlapEtlTask.html:
```diff
@@ -0,0 +1,294 @@
+                                    <label for="maxNumber" class="control-label">Max Items in Row</label>
+                                    <div class="flex-grow" data-bind="validationElement: maxNumberOfItemsInRowGroup">
+                                        <div class="input-group">
+                                            <input type="number" min="1" class="form-control" placeholder="Enter a number (or leave empty for null)" id="maxNumber"
```
leave empty for null? What null means?

ml054 replied:
can we avoid id="maxNumber"?

Danielle9897 replied:
advanced removed

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editOlapEtlTask.html:
```diff
@@ -0,0 +1,294 @@
+                <div class="panel" data-bind="visible: showEditOlapTableArea()">
+                    <div class="padding padding-sm flex-horizontal item" data-bind="with: editedOlapTableSandbox()">
+                        <div class="flex-grow" data-bind="validationElement: tableName"> 
+                            <!-- todo 1: why not red on the FIRST Time ? only after saving at least once.. -->
```
TODO

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editSqlEtlTask.html:
```diff
@@ -312,7 +314,7 @@ <h3 class="text-center text-muted" data-bind="css: { 'text-danger': !transformat
                                 <label class="control-label">Document ID</label>
                                 <div class="flex-grow" data-bind="validationElement: documentId">
-                                    <input type="text" id="documentId" data-bind="textInput: documentId" class="form-control" placeholder="Document ID"
+                                    <input type="text" id="documentId" data-bind="textInput: documentId" class="form-control" placeholder="Document ID" autocomplete="off"
```
don't we already have autocomplete off on entire form?

Danielle9897 replied:
there is autocomplete on entire 'form' which is on the left side of screen
this was missing on the right side

aviv86 committed:
RavenDB-15742 : support no partition, multiple partitions, change script syntax

aviv86 committed:
RavenDB-15742 : change RunFrequency from TimeSpan to cron expression

aviv86 committed:
RavenDB-15742 : various changes
- add ETL.OLAP.MaxNumberOfExtractedDocuments ETL configuration and use it in CanContinueBatch
- remove MaxNumberOfItemsInRowGroup
- remove PartitionColumn from OlapTables
- add CustomPrefix to remote path if needed + test
- minor changes in S3 tests

aviv86 committed:
RavenDB-15742 : server-wide commands

aviv86 committed:
RavenDB-16310 OLAP ETL : Connection Strings View

aviv86 committed:
RavenDB-15742 : S3-tests minor fixes

Danielle9897 committed:
RavenDB-16310 OLAP ETL : rebase + fix studio for latest api

aviv86 committed:
RavenDB-15742 : support Azure as OLAP destination + tests

Danielle9897 committed:
RavenDB-16310 OLAP ETL : fix studio for latest api - part 2

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/OLAP/OlapConnectionString.cs:
```diff
@@ -0,0 +1,63 @@
+            if (S3Settings != null)
+            {
+                if (S3Settings.HasSettings() == false)
+                    errors.Add($"{nameof(S3Settings)} has no valid setting");
```
Shouldn't we tell user exactly what is not valid?

arekpalinski commented on src/Raven.Client/Documents/Operations/ETL/OLAP/OlapConnectionString.cs:
```diff
@@ -0,0 +1,63 @@
+            }
+
+            if (LocalSettings.HasSettings() == false)
+                errors.Add($"{nameof(LocalSettings)} has no valid setting");
```
Same here - can we provide a user with concrete field error?

arekpalinski commented on src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:
```diff
@@ -40,20 +40,22 @@ public enum DatabaseRecordItemType
+        TimeSeries = 1 << 16,
+        DocumentsCompression = 1 << 17,
+        Analyzers = 1 << 18,
+        LockMode = 1 << 19
```
Just to make sure - we don't store those enums anywhere (e.g. in an export file)? Asking about because you modified existing values so I wanna be sure it doesn't break the import compatibility of older .ravendump files?

ppekrol replied:
To stay on the safe side, let's revert this to original values and add OlapEtl and OlapConnectionStrings at the end

arekpalinski commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1311,6 +1323,12 @@ public void AssertCanAddSqlEtl()
         
+        public void AssertCanAddOlapEtl()
+        {
+            // RavenDB-16242
```
https://issues.hibernatingrhinos.com/issue/RavenDB-16242 is fixed

arekpalinski commented on src/Raven.Server/Config/Categories/EtlConfiguration.cs:
```diff
@@ -41,5 +41,10 @@ public class EtlConfiguration : ConfigurationCategory
         public Size? MaxBatchSize { get; protected set; }
+
+        [Description("Max number of extracted documents in OLAP ETL batch")]
+        [DefaultValue(50 * 1024)]
```
Maybe let's use power of 2 as the default - 65536?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlapDocumentTransformer.cs:
```diff
@@ -0,0 +1,245 @@
+            for (var i = 0; i < result.Count; i++)
+            {
+                result.GetPropertyByIndex(i, ref prop);
+                props.Add(new SqlColumn
```
SqlColumn? Maybe create a dedicated class?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlapDocumentTransformer.cs:
```diff
@@ -0,0 +1,245 @@
+        }
+
+        protected override void AddLoadedAttachment(JsValue reference, string name, Attachment attachment)
+        {
```
Shouldn't we throw `NotSupportedException` here?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlapDocumentTransformer.cs:
```diff
@@ -0,0 +1,245 @@
+        }
+
+
+        private JsValue LoadToS3FunctionTranslator(string name, JsValue[] args)
```
Why does it have "S3" in the name? I think it's not s3 specific, right?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlapDocumentTransformer.cs:
```diff
@@ -0,0 +1,245 @@
+        public override void Transform(ToOlapItem item, EtlStatsScope stats, EtlProcessState state)
+        {
+            _stats = stats;
+            if (item.IsDelete)
```
Given that we don't do anything with deleted docs, do we actually extract tombstones? If yes, then can we skip it completely?

arekpalinski replied:
I found it later that we do skip extraction of tombstones. Then please add here a comment that will explain that actually this must never happen. You can even add Debug.Assert here

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlapEtlMetricsCountersManager.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Raven.Server.Documents.ETL.Providers.OLAP
+{
+    // RavenDB-16309
```
Just a reminder about RavenDB-16309

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlaptEtl.cs:
```diff
@@ -0,0 +1,250 @@
+            _timer?.Dispose();
+
+            if (Logger.IsOperationsEnabled)
+                Logger.Operations($"Next OLAP ETL run is in {nextRun.TimeSpan.TotalMinutes} minutes.");
```
1. Shouldn't we add `Name` to the message?
2. How often is it going to be called? Asking to make sure it won't flood the logs (since it's on operations level)

aviv86 replied:
Added `Name` to the message. 
This will be called once a minute at most

arekpalinski replied:
Once a minute is quite often actually. For me it looks rather like debug logs so I'd change it to `Information` level. What do you think?

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/OLAP/OlaptEtl.cs:
```diff
@@ -0,0 +1,250 @@
+
+        private static void ProgressNotification(IOperationProgress progress)
+        {
+            // todo RavenDB-16341
```
Just reminder about TODO

arekpalinski commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -143,6 +143,7 @@
     <PackageReference Include="NuGet.Protocol" Version="5.8.1" />
     <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="3.21.1" />
     <PackageReference Include="Portable.BouncyCastle" Version="1.8.10" />
+    <PackageReference Include="Parquet.Net" Version="3.8.4" />
```
Make sure we're using the latest `Parquet.Net`

aviv86 replied:
updated to 3.8.6

arekpalinski commented on src/Raven.Server/Web/System/OngoingTasksHandler.cs:
```diff
@@ -847,6 +863,34 @@ private IEnumerable<OngoingTask> CollectEtlTasks(DatabaseRecord databaseRecord,
+                            NodeUrl = clusterTopology.GetUrlFromTag(tag)
+                        },
+                        ConnectionStringName = olapEtl.ConnectionStringName,
+                        //Destination = olapEtl.GetDestination(), // todo ... waiting for RavenDB-16435  
```
TODO reminder

grisha-kotler committed:
RavenDB-15742 - add olapEtl license attribute

grisha-kotler committed:
RavenDB-15742 - EnsureTaskNameIsNotUsed for OLAP ETL

aviv86 committed:
RavenDB-15742 : ParquetTransformedItems : fix lazy number handling, and type changes in numbers (e.g. 'long' change to 'double')

Danielle9897 committed:
RavenDB-16310 OLAP ETL : Add Azure destination in Studio

aviv86 committed:
RavenDB-15742 : - OLAP ETL : avoid updating etl state when we didn't do any work
- ParquetTransformedItems : minor fixes in type-change handling

aviv86 committed:
RavenDB-15742 : encode special chars in order to ensure safe file name

aviv86 committed:
RavenDB-15742 : various changes
- ParquetTransformedItems : remove DefaultMaxItemsInGroup, support specifying type in script e.g. { Value : x, Type : t}, refactoring
- remove KeepFilesOnDisk from OlapEtlConfiguration
- change DefaultPartitionColumnName
- add tests

aviv86 committed:
RavenDB-15742 : change from CustomPrefix to CustomField, support partitioning by it via script

aviv86 committed:
RavenDB-15742 : address PR comments, fix flaky test ShouldRespectEtlRunFrequency, fix failing tests

ppekrol committed:
RavenDB-15742 support for loadTo(name, key, obj)

Danielle9897 committed:
RavenDB-16310 OLAP ETL : Apply latest api to Studio

ppekrol committed:
RavenDB-15742 added name uniqueness test

ppekrol committed:
RavenDB-15742 moved name uniqueness test to a dedicated class

ppekrol committed:
RavenDB-15742 rename

ppekrol committed:
RavenDB-15742 smuggler support for OLAP ETL

Danielle9897 committed:
RavenDB-16310 OLAP ETL : Unify script prefix name

ppekrol committed:
RavenDB-15742 added olap restore test

Danielle9897 committed:
RavenDB-16310 OLAP ETL : A fix for the tasks list view

aviv86 committed:
RavenDB-15742 : change syntax of partitionBy()

ppekrol commented on src/Raven.Server/Config/Categories/EtlConfiguration.cs:
```diff
@@ -41,5 +41,10 @@ public class EtlConfiguration : ConfigurationCategory
         public Size? MaxBatchSize { get; protected set; }
+
+        [Description("Max number of extracted documents in OLAP ETL batch")]
+        [DefaultValue(65536)]
```
Sometimes it is just easier to write `64 * 1024`

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/OLAP/ParquetTransformedItems.cs:
```diff
@@ -0,0 +1,712 @@
+                            array = _boolArr ??= new bool[data.Count];
+                            break;
+                        case DataType.Byte:
+                            array = _byteArr ??= new byte[data.Count];
```
Should we use ArrayPool for those?

ppekrol commented on src/Raven.Server/Json/JsonDeserializationServer.cs:
```diff
@@ -44,7 +44,11 @@
 using Raven.Server.Web.Studio;
 using Raven.Server.Web.System;
 using Sparrow.Json;
+using Raven.Server.ServerWide;
```
Let's fix those warnings

ppekrol commented on src/Raven.Server/ServerWide/Commands/ConnectionStrings/PutConnectionStringCommand.cs:
```diff
@@ -70,4 +71,33 @@ public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```
Remove `try catch`

ppekrol committed:
RavenDB-15742 fixed DatabaseItemType and DatabaseRecordItemType

ppekrol committed:
RavenDB-15742 cleanup

ppekrol committed:
RavenDB-15742 fixing test

ppekrol committed:
RavenDB-15742 added load/file and load/upload scope

ppekrol committed:
RavenDB-15742 do not add load/upload scope if there are no uploads
